core/blockchain/blockchain_utils.py

```diff
@@ -0,0 +1,161 @@
+import json
+import logging
+import requests
+import time
+from typing import Callable
+
+from core.utils.tx_utils import TxEnum, Transaction
+
+
+# TODO: does this need to be here?
+logging.basicConfig(level=logging.DEBUG,
+    format='[BlockchainGateway] %(message)s')
+
+
+##############################################################################
+###                              REQUIRE IPFS                              ###
+##############################################################################
+
+def upload(client: object, value: dict) -> str:
+    """
+    Provided any Python object, store it on IPFS and then upload the hash that
+    will be uploaded to the blockchain as a value
+    """
+    # assert TxEnum.KEY.name in value
+    # assert TxEnum.CONTENT.name in value
+    ipfs_hash = content_to_ipfs(client, value)
+    return str(ipfs_hash)
+
+def download(client: object, key: str, state: list) -> list:
+    """
+    Provided an on-chain key, retrieve the value from local state and retrieve
+    the Python object from IPFS
+    TODO: implement a better way to parse through state list
+    """
+    relevant_txs = list(
+        map(lambda tx: ipfs_to_content(client, tx.get(TxEnum.CONTENT.name)),
+        filter(lambda tx: tx.get(TxEnum.KEY.name) == key, state)))
+    return relevant_txs
+
+def ipfs_to_content(client: object, ipfs_hash: str) -> object:
+    """
+    Helper function to retrieve a Python object from an IPFS hash
+    """
+    return client.get_json(ipfs_hash)
+
+def content_to_ipfs(client: object, content: dict) -> str:
+    """
+    Helper function to deploy a Python object onto IPFS, returns an IPFS hash
+    """
+    return client.add_json(content)
+
+##############################################################################
+###                                 DIFFS                                  ###
+##############################################################################
+
+def get_diffs(global_state: list, local_state: list) -> list:
+    """
+    Return list of transactions that are present in `global_state` but not in
+    `local_state`
+    """
+    len_local_state = len(local_state)
+    return global_state[len_local_state:]
```

Comment:
Nitpick: `len(local_state)` is self-documenting, you can honestly just replace with: `return global_state[len(local_state):]`